### PR TITLE
Add the RPG-Styled HP and MP Counter patch

### PR DESCRIPTION
**Pull request by ExE-Boss**
Since the original RPG-Styled HP and MP Counter patch which is available on SMW Central still hasn’t been made SA-1 compatible and I’ve long since been maintaining my own fork to support the changes that I’ve made for my own hack, which I’ve showcased on the SMW Central Discord server a few times.

I’ve decided to merge my own improvements into the patch, update it to match the code style of this repository and upload it here (see #1).

### Depends on:
- #2

## Changes
- Optimised a lot of code
- Added SA-1 support
- Added Sprite Status Bar to the Bowser battle, can be turned off if another patch already handles this
- Added support for VWF Dialogues and Free $7F4000
- Made it so that spin-jumping doesn’t auto-shoot fireballs everywhere
- Fireballs can now only be shot using the X button if they cost MP, this is to prevent wasteful usage of MP
- Knockback is now customisable
- Added support for more than 2 fireballs (requires the [Shoot More Fireballs](https://www.smwcentral.net/?p=section&a=details&id=4469) patch)
- The Player head can now be moved around or removed altogether
- The status bar can now display MP values of up to 255, but the rest of the code still only supports values up to 99

### CODE REVIEW

**Comment by RPGHacker** [comment]
And thanks again for your contribution! :)
That sounds like some nice new features. Not much left to do on the Asar manual I'm currently working on. Once that's done, I'll try to take some time so that I can merge all of your pull requests.

File: patches/hp_counter_patch/hp_counter_patch.asm [hunk 1]
```diff
@@ -0,0 +1,1366 @@
...
+?TwoDigit:
+	stx get_status_ram(<x>+1,<y>)
+endmacro
+
+
+; Used for the sprite status bar
+macro BowserBattle_SBTile(dest,x,y,tile,size)
+	ldy.w #<dest>
+	lda.b #<size>
+	sta.w remap_ram($0420),y
```
Comment (ExE-Boss) [inline patches/hp_counter_patch/hp_counter_patch.asm]: This should use hardcoded addressing.

File: patches/hp_counter_patch/hp_counter_patch.asm [hunk 2]
```diff
@@ -0,0 +1,1366 @@
...
+	tya
+	asl #$2
+	tay
+	lda.b #<x>
+	sta.w remap_ram($0200),y
```
Comment (ExE-Boss) [inline patches/hp_counter_patch/hp_counter_patch.asm]: ditto

File: patches/hp_counter_patch/hp_counter_patch.asm [hunk 3]
```diff
@@ -0,0 +1,1366 @@
...
+	lda.b #<y>
+	sta.w remap_ram($0201),y
```
Comment (ExE-Boss) [inline patches/hp_counter_patch/hp_counter_patch.asm]: ditto

File: patches/hp_counter_patch/hp_counter_patch.asm [hunk 4]
```diff
@@ -0,0 +1,1366 @@
...
+	lda.b #<tile>+$C0	;\
+	sta.w remap_ram($0202),y	; | Enforce that bowser’s battle mode always uses lines 4-5 of SP2 (underwater BG),
```
Comment (ExE-Boss) [inline patches/hp_counter_patch/hp_counter_patch.asm]: ditto

File: patches/hp_counter_patch/hp_counter_patch.asm [hunk 5]
```diff
@@ -0,0 +1,1366 @@
...
+	lda.b #$30	; | Which are the only unused tiles in this battle mode.
+	sta.w remap_ram($0203),y	;/
```
Comment (ExE-Boss) [inline patches/hp_counter_patch/hp_counter_patch.asm]: ditto

**Comment by ExE-Boss** [comment]
@RPGHacker Now that the **VWF Dialogues Patch v1.2** [is officially released](https://www.smwcentral.net/?p=section&a=details&id=17485), you should be able to review this.

**Pull request merged**